### PR TITLE
Keycloak: add client_rolemapping management

### PR DESCRIPTION
**Pull request by Gaetan2907**
##### SUMMARY
This PR add a module to manage keycloak client_rolemapping, like keycloak_client manage clients

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
`keycloak_client_rolemapping.py`

### CODE REVIEW

File: plugins/module_utils/identity/keycloak/keycloak.py [hunk 1]
```diff
@@ -377,7 +381,7 @@ def update_client(self, id, clientrep, realm="master"):
     def create_client(self, clientrep, realm="master"):
         """ Create a client in keycloak
         :param clientrep: Client representation of client to be created. Must at least contain field clientId
-        :param realm: realm for client to be created
+        :param realm: realm for client to be created.
```
Comment (felixfontein) [inline plugins/module_utils/identity/keycloak/keycloak.py:384]: Out of curiosity: why change this line, but not the one above, where a trailing period is also missing?

File: plugins/module_utils/identity/keycloak/keycloak.py [hunk 2]
```diff
@@ -405,6 +409,125 @@ def delete_client(self, id, realm="master"):
...
+        """ Fetch the available role of a client in a specified goup on the Keycloak server.
+
+        :param gid: ID of the group from which to obtain the rolemappings.
+        :param cid: ID of the client from which to obtain the rolemappings.
+        :param realm: Realm from which to obtain the rolemappings.
+        :return: The rollemappings of specified group and client of the realm (default "master").
+        """
+        available_rolemappings_url = URL_CLIENT_ROLEMAPPINGS_AVAILABLE.format(url=self.baseurl, realm=realm, id=gid, client=cid)
+        try:
+            # ret_val = open_url(available_rolemappings_url, method="GET", headers=self.restheaders,
```
Comment (felixfontein) [inline plugins/module_utils/identity/keycloak/keycloak.py:473]: I would remove the commented out code.

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: keycloak_client_rolemapping
+
+short_description: Allows administration of Keycloak client_scopes via Keycloak API
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:13]:
```suggestion
short_description: Allows administration of Keycloak client_scopes with the Keycloak API
version_added: 3.5.0
```

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+description:
+    - This module allows you to add, remove or modify Keycloak client_rolemapping via the Keycloak REST API.
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:16]:
```suggestion
    - This module allows you to add, remove or modify Keycloak client_rolemapping with the Keycloak REST API.
```

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        type: str
+        description:
+            - Id of the group to be mapped
+            - This parameter is not required for updating or deleting the rolemapping but
+              providing it will reduce the number of API calls required.
+
+    client_id:
+        type: str
+        description:
+            - Name of the client to be mapped (different than cid).
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:67]:
```suggestion
            - Name of the client to be mapped (different than I(cid)).
```

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+            - This parameter is required (can be replaced by cid for less API call).
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:68]:
```suggestion
            - This parameter is required (can be replaced by I(cid) for less API calls).
```
Also: if it is required, why is there no `required: true`?

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+    cid:
+        type: str
+        description:
+            - Id of the client to be mapped
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:73]:
```suggestion
            - Id of the client to be mapped.
```

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+            - This parameter is not required for updating or deleting the rolemapping but
+              providing it will reduce the number of API calls required.
+
+    roles:
+        description:
+            - Roles to be mapped to the group
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:79]:
```suggestion
            - Roles to be mapped to the group.
```

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+
+RETURN = '''
+msg:
+  description: Message as to what action was taken
+  returned: always
+  type: str
+  sample: "Role role1 assigned to group group1."
+
+proposed:
+    description: role_representation representation of proposed changes to client_scope
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:166]:
```suggestion
    description: role_representation representation of proposed changes to client_scope.
```

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+    returned: always
+    type: dict
+    sample: {
+      clientId: "test"
+    }
+existing:
+    description: role_representation representation of existing role_representation (sample is truncated)
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:173]:
```suggestion
    description:
      - role_representation representation of existing role_representation.
      - The sample is truncated.
```

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+    returned: always
+    type: dict
+    sample: {
+        "adminUrl": "http://www.example.com/admin_url",
+        "attributes": {
+            "request.object.signature.alg": "RS256",
+        }
+    }
+end_state:
+    description: role_representation representation of role_representation after module execution (sample is truncated)
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:183]:
```suggestion
    description:
      - role_representation representation of role_representation after module execution.
      - The sample is truncated.
```

**Review by felixfontein — COMMENTED** [review]
Where is check mode implemented?

File: plugins/module_utils/identity/keycloak/keycloak.py (hunk 1 above)
Comment (Gaetan2907) [reply plugins/module_utils/identity/keycloak/keycloak.py:384]: It was an omission, I will also add it for the above line.

Comment (Gaetan2907) [reply plugins/module_utils/identity/keycloak/keycloak.py:384]: Fix with commit ccd1971

File: plugins/module_utils/identity/keycloak/keycloak.py (hunk 2 above)
Comment (Gaetan2907) [reply plugins/module_utils/identity/keycloak/keycloak.py]: Fix with commit ccd1971

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py (hunk 5 above)
Comment (Gaetan2907) [reply plugins/modules/identity/keycloak/keycloak_client_rolemapping.py]: Fix with 22c76d9

**Comment by Gaetan2907** [comment]
@felixfontein Thanks for your review !

**Comment by Gaetan2907** [comment]
> Where is check mode implemented?

You're right diff mode is there but check mode is missing... 
I'll add it asap

**Comment by felixfontein** [comment]
Please note that we added a new sanity test which requires a BOTMETA entry for every new plugin and module. See [this section on adding new plugins and modules](https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#creating-new-modules-or-plugins) for details.

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py (hunk 6 above)
Comment (felixfontein) [reply plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:69]: @Gaetan2907 please don't forget this one :)

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+            assigned_roles_after = kc.get_client_composite_rolemappings(gid, cid, realm=realm)
+            result['end_state'] = assigned_roles_after
+            module.exit_json(**result)
+    # Do nothing
+    else:
+        result['changed'] = False
+        result['msg'] = 'Nothing to do, roles %s are correctly mapped with group %s.' % (roles, group_name)
+        module.exit_json(**result)
+
+    module.exit_json(**result)
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:345]:
```suggestion
```
This should not be needed, when looking at the above code?

File: plugins/modules/identity/keycloak/keycloak_client_rolemapping.py [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+            assigned_roles_after = kc.get_client_composite_rolemappings(gid, cid, realm=realm)
+            result['end_state'] = assigned_roles_after
+            module.exit_json(**result)
+    # Do nothing
+    else:
+        result['changed'] = False
+        result['msg'] = 'Nothing to do, roles %s are correctly mapped with group %s.' % (roles, group_name)
+        module.exit_json(**result)
+
+
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_client_rolemapping.py:345]:
```suggestion
```

**Review by felixfontein — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@Gaetan2907 thanks a lot for contributing this module!

**Pull request merged**